testhelpers/server_factory.go

```diff
@@ -0,0 +1,45 @@
+package testhelpers
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+type HTTPTestResponse struct {
+	Code int
+	Body string
+}
+
+type HTTPTestRequest struct {
+	Path   string
+	Method string
+	Body   string
+}
+
+func BuildHTTPTestServer(request *HTTPTestRequest, response *HTTPTestResponse, requestLabel TestRequestLabel) *httptest.Server {
+	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		defer GinkgoRecover()
+
+		TestRequestTracker <- requestLabel
+
+		if request.Path != "" {
+			Expect(r.URL.Path).To(Equal(request.Path))
+		}
+		if request.Method != "" {
+			Expect(r.Method).To(Equal(request.Method))
+		}
+
+		body, err := ReadHTTPBody(r.Body)
+		Expect(err).To(BeNil())
+		if request.Body != "" {
+			Expect(body).To(MatchJSON(request.Body))
+		}
+
+		w.WriteHeader(response.Code)
+		fmt.Fprintln(w, response.Body)
+	}))
```

Comment:
This feel a bit like it's coupling the construction of a test server with the expectations of what requests will hit it.  As it stands, it's not possible to construct a test server that handles multiple different requests.

It may be better to construct a test server that records the requests made, and then allows inspection on them (eg in the router test suite[1])

An alternative approach would be to construct a test server that could then have expectations set on it - as done in the [contentstore client tests](https://github.com/alphagov/publishing-api/blob/master/contentstore/client_test.go#L36-L43).

[1] https://github.com/alphagov/router/blob/master/integration_tests/proxy_function_test.go#L149-L160 and https://github.com/alphagov/router/blob/master/integration_tests/backend_helpers.go#L44-L49